File: F#/XamarinStore.iOS/Views/ImageButton.fs
namespace XamarinStore

open System
open MonoTouch.UIKit
open System.Drawing
open MonoTouch.CoreGraphics

type ImageButton() as this =
    inherit UIControl(new RectangleF(0.0f,0.0f,250.0f, 50.0f))

    let label = new UILabel(TextColor = this.TintColor)
    let imageView = new UIImageView()

    do this.AddSubview imageView
       this.AddSubview label

       this.Layer.BorderColor <- this.TintColor.CGColor
       this.Layer.BorderWidth <- 1.0f
       this.Layer.CornerRadius <- 5.0f

    member val ImageView = imageView with get

    override this.TintColorDidChange () =
        base.TintColorDidChange ()
        this.Layer.BorderColor <- this.TintColor.CGColor
        label.TextColor <- this.TintColor

    override this.LayoutSubviews () =
        base.LayoutSubviews ()

        let padding = 10.0f
        let imageSize = this.ImageView.SizeThatFits this.Bounds.Size

        let mutable availableWidth = this.Bounds.Width - padding * 3.0f - imageSize.Width
        let stringSize = label.SizeThatFits (new System.Drawing.SizeF (availableWidth, this.Bounds.Height - padding * 2.0f))

        let availableWidth = this.Bounds.Width - stringSize.Width - imageSize.Width
       
        let x = availableWidth / 2.0f

        let mutable frame = new RectangleF (new PointF (x, this.Bounds.GetMidY () - imageSize.Height / 2.0f), imageSize)
        imageView.Frame <- frame

        frame.X <- frame.Right + if imageSize.Width > 0.0f then padding else 0.0f
        frame.Size <- stringSize
        frame.Height <- this.Bounds.Height
        frame.Y <- 0.0f
        label.Frame <- frame

    member this.Text 
        with get () = label.Text
        and set value = label.Text <- value

    member this.Image 
        with get () = imageView.Image
        and set value = imageView.Image <- value

    member this.Font 
        with get () = label.Font
        and set value = label.Font <- value

    override this.TouchesBegan (touches, evt) =
        base.TouchesBegan (touches, evt)
        this.TintAdjustmentMode <- UIViewTintAdjustmentMode.Dimmed

    override this.TouchesEnded (touches, evt) =
        base.TouchesEnded (touches, evt)
        this.TintAdjustmentMode <- UIViewTintAdjustmentMode.Automatic

    override this.TouchesCancelled (touches, evt) =
        base.TouchesCancelled (touches, evt)
        this.TintAdjustmentMode <- UIViewTintAdjustmentMode.Automatic

File: C#/XamarinStore.iOS/Views/ImageButton.cs
using System;
using MonoTouch.UIKit;
using System.Drawing;
using MonoTouch.CoreGraphics;

namespace XamarinStore
{
	public class ImageButton : UIControl
	{
		public readonly UIImageView ImageView;
		UILabel label;
		public ImageButton () : base(new RectangleF(0,0,250,50))
		{
			ImageView = new UIImageView ();
			AddSubview (ImageView);

			label = new UILabel {
				TextColor = TintColor,
			};
			AddSubview (label);

			Layer.BorderColor = TintColor.CGColor;
			Layer.BorderWidth = 1f;
			Layer.CornerRadius = 5f;
		}
		public override void TintColorDidChange ()
		{
			base.TintColorDidChange ();
			Layer.BorderColor = TintColor.CGColor;
			label.TextColor = TintColor;
		}
		public override void LayoutSubviews ()
		{
			base.LayoutSubviews ();

			const float padding = 10f;
			var imageSize = ImageView.SizeThatFits (Bounds.Size);

			var availableWidth = Bounds.Width - padding * 3 - imageSize.Width;
			var stringSize = label.SizeThatFits (new System.Drawing.SizeF (availableWidth, Bounds.Height - padding * 2));

			availableWidth = Bounds.Width ;
			availableWidth -= stringSize.Width;
			availableWidth -= imageSize.Width;

			var x = availableWidth / 2;

			var frame = new RectangleF (new PointF (x, Bounds.GetMidY () - imageSize.Height / 2), imageSize);
			ImageView.Frame = frame;

			frame.X = frame.Right + (imageSize.Width > 0 ? padding : 0);
			frame.Size = stringSize;
			frame.Height = Bounds.Height;
			frame.Y = 0;
			label.Frame = frame;
		}

		public string Text {
			get{ return label.Text; }
			set{ label.Text = value; }
		}

		public UIImage Image {
			get{ return ImageView.Image; }
			set{ 
				ImageView.Image = value;
			 }
		}

		public UIFont Font {
			get{ return label.Font; }
			set{ label.Font = value; }
		}

		public override void TouchesBegan (MonoTouch.Foundation.NSSet touches, UIEvent evt)
		{
			base.TouchesBegan (touches, evt);
			TintAdjustmentMode = UIViewTintAdjustmentMode.Dimmed;
		}

		public override void TouchesEnded (MonoTouch.Foundation.NSSet touches, UIEvent evt)
		{
			base.TouchesEnded (touches, evt);
			TintAdjustmentMode = UIViewTintAdjustmentMode.Automatic;
		}
		public override void TouchesCancelled (MonoTouch.Foundation.NSSet touches, UIEvent evt)
		{
			base.TouchesCancelled (touches, evt);
			TintAdjustmentMode = UIViewTintAdjustmentMode.Automatic;
		}

	}
}

